### 문제1

1991년 일반사회조사에서 조사된 다음의 4X4 분할표는 흑인들에 대해서 직업만족도와 수입에 대해서 조사한 결과를 보여준다.

|수입|직업만족도<br>매우 불만족|직업만족도<br>약간 만족|직업만족도<br>적절히 만족|직업만족도<br>매우 만족|
|:-:|:-:|:-:|:-:|:-:|
|<5000|2|4|13|3|
|5000 ~ 15000|2|6|22|4|
|15000 ~ 25000|0|1|15|8|
|>25000|0|3|13|8|

a. <img src="https://latex.codecogs.com/svg.image?\chi^2"> 검정을 이용하여 직업만족도와 수입 간의 독립성검정을 수행하라. 검정 결과를 해석하고 이 자료에 대해서 이 검정이 갖고 있는 문제점을 설명하라. 표준화잔차를 구하라. 어떠한 패턴을 보이고 있는가?

In [62]:
# 데이터 입력
data = matrix(c(2, 4, 13, 3,
                2, 6, 22, 4,
                0, 1, 15, 8,
                0, 3, 13, 8),
                nrow=4, byrow=TRUE)

colnames(data) = c("매우불만족", "약간만족", "적절히만족", "매우만족")
rownames(data) = c("<5000", "5000~15000", "15000~25000", ">25000")

In [20]:
# 카이제곱 검정을 이용하여 직업만족도와 수입 간의 독립성검정 수행
result = chisq.test(data)
result

Warning message in chisq.test(data):
"Chi-squared approximation may be incorrect"



	Pearson's Chi-squared test

data:  data
X-squared = 11.524, df = 9, p-value = 0.2415


< 카이제곱 검정 결과 >
chisq.test 함수를 통해 카이제곱검정을 실시한 결과,  
검정 통계량이 11.524, 자유도는 9, p값은 0.2415 이다.  

자유도 df = (I-1)(J-1) 이므로 (4-1)×(4-1) = 9 이다.  
p값은 1-pchisq(11.524, df=9) 명령어를 통해 0.2415임을 확인할 수 있다.  

In [21]:
p_value = 1-pchisq(11.524, df=9)
p_value

[1] 0.2414925

< 검정결과에 대한 해석 >

귀무가설은 "수입(독립변수)과 직업만족도(반응변수)는 서로 독립이다" 이며,  
대립가설은 "수입(독립변수)과 직업만족도(반응변수)는 서로 독립적이지 않다." 이다.

검정 결과 p값이 0.2415로서 0.05보다 크므로 귀무가설을 기각할 수 없는 것으로 해석할 수 있다.

< 이 검정이 갖고 있는 문제점 >

다만, 행과 열의 각각 4개씩 존재하는 분할표에서 전체 관측치가 104개라면 평균적으로 하나의 셀에 6.5개의 관측이 존재할 수 있는데, 각 셀마다 관측치 수의 편차가 크다는 점을 고려하면 검정결과가 왜곡될 우려가 있다고 판단된다.

In [22]:
# 표준화잔차 구하기
result$stdres

매우불만족 약간만족   적절히만족 매우만족 
<5000        1.4406207  0.7305280 -0.1606257 -1.079180
5000~15000   0.7525419  0.8715747  0.6005081 -1.772568
15000~25000 -1.1171392 -1.5211385  0.2198088  1.509794
>25000      -1.1171392 -0.1573592 -0.7326960  1.509794

< 표준화잔차의 패턴 >
표준화잔차의 패턴을 살펴보면  
첫째, 수입이 낮으면 매우불만족, 약간만족이 기대값보다 많이 나타나고, 수입이 높으면 매우불만족, 약간만족이 기대값보다 적게 나타난다.  
둘째, 수입이 낮으면 적절히만족, 매우만족이 기대값보다 대체로 적게 나타나고, 수입이 높으면 적절히만족, 매우만족이 기대값보다 대체로 많이 나타난다.

b. 수입에 대해서는 (3, 10, 20, 35) 점수를 사용하고 직업만족도에 대해서는 (1, 3, 4, 5) 점수를 사용하여 변수들을 질적인 형태로 간주한 후에 검정을 수행하라. 왜 이 검정 결과가 (a)와 다른지 설명하라.

![SAS 코드](02강%20과제1-2.png)

![SAS 결과](02강%20과제1-2_결과.png)

< 점수를 반영하여 질적인 형태로 간주한 변수의 검정 결과 >
(a)에서 실시한 카이제곱 검정에 따른 p값은 0.2415 였는데, (b)에서 점수를 부여한 Mentel-Haenszel 카이제곱 검정에 따른 p 값은 0.0079이다.
즉, Mentel-Haenszel 카이제곱 검정에 따르면 귀무가설을 기각할 수 있으며, 수입과 만족도 사이에는 연관성이 존재한다.

< 검정 결과가 a와 다른 이유 >
데이터에 점수를 반영하면 데이터는 단순히 범주형이 아니라 순서형 정보를 포함하게 된다.
점수를 반영한 데이터는 변수의 순서와 강도를 추가로 반영하게 되며, 이는 검정통계량과 p값에 큰 영향을 미치며, 이에 따라 변수간의 관계에도 영향을 미치는 것으로 보인다.

In [71]:
library(MASS)

# 데이터 입력
# X는 독립 변수, Y는 순서형 반응 변수
data <- data.frame(
  X = factor(c(3, 3, 3, 3, 10, 10, 10, 10, 20, 20, 20, 20, 35, 35, 35, 35)),
  Y = factor(c(1, 3, 4, 5, 1, 3, 4, 5, 1, 3, 4, 5, 1, 3, 4, 5), 
             levels = c(1, 3, 4, 5), ordered = TRUE),
  count = c(2, 12, 52, 15, 2, 6, 22, 4, 0, 1, 15, 8, 0, 3, 13, 8)
)

# 다항 로지스틱 회귀 모델 적합
model <- polr(Y ~ X, data = data, weights = count, Hess = TRUE)

# 모델 요약
summary(model)


Call:
polr(formula = Y ~ X, data = data, weights = count, Hess = TRUE)

Coefficients:
      Value Std. Error t value
X10 -0.4632     0.4178  -1.109
X20  0.9145     0.4632   1.975
X35  0.7026     0.4761   1.476

Intercepts:
    Value   Std. Error t value
1|3 -3.6306  0.5312    -6.8352
3|4 -1.5871  0.2668    -5.9489
4|5  1.5018  0.2621     5.7294

Residual Deviance: 312.4581 
AIC: 324.4581 

In [70]:
# 필요한 패키지 설치 및 로드
if (!requireNamespace("MASS", quietly = TRUE)) {
  install.packages("MASS")
}
library(MASS)

# 데이터 입력
data <- data.frame(
  X = factor(c(3, 3, 3, 3, 10, 10, 10, 10, 20, 20, 20, 20, 35, 35, 35, 35)),
  Y = factor(c(1, 3, 4, 5, 1, 3, 4, 5, 1, 3, 4, 5, 1, 3, 4, 5)),
  count = c(2, 12, 52, 15, 2, 6, 22, 4, 0, 1, 15, 8, 0, 3, 13, 8)
)

# 각 (X, Y) 조합의 비율을 계산하여 다항 로지스틱 회귀 분석을 위한 데이터 준비
# Y를 순서형 반응변수로 설정
data$Y <- factor(data$Y, levels = sort(unique(data$Y)))

# 다항 로지스틱 회귀 분석 모델 적합
model <- polr(Y ~ X, data = data, weights = count, Hess = TRUE)

# 모델 요약
summary(model)


Call:
polr(formula = Y ~ X, data = data, weights = count, Hess = TRUE)

Coefficients:
      Value Std. Error t value
X10 -0.4632     0.4178  -1.109
X20  0.9145     0.4632   1.975
X35  0.7026     0.4761   1.476

Intercepts:
    Value   Std. Error t value
1|3 -3.6306  0.5312    -6.8352
3|4 -1.5871  0.2668    -5.9489
4|5  1.5018  0.2621     5.7294

Residual Deviance: 312.4581 
AIC: 324.4581 

In [69]:
library(dplyr)

# 데이터 입력
data <- data.frame(
  X = c(3, 3, 3, 3, 10, 10, 10, 10, 20, 20, 20, 20, 35, 35, 35, 35),
  Y = c(1, 3, 4, 5, 1, 3, 4, 5, 1, 3, 4, 5, 1, 3, 4, 5),
  count = c(2, 4, 13, 3, 2, 6, 22, 4, 0, 1, 15, 8, 0, 3, 13, 8)
)

# 교차표 생성
table_data <- xtabs(count ~ X + Y, data = data)
table_data

# 카이제곱 검정 수행
chi_sq_test <- chisq.test(table_data)

# 결과 출력
print(chi_sq_test)


    Y
X     1  3  4  5
  3   2  4 13  3
  10  2  6 22  4
  20  0  1 15  8
  35  0  3 13  8

Warning message in chisq.test(table_data):
"Chi-squared approximation may be incorrect"



	Pearson's Chi-squared test

data:  table_data
X-squared = 11.524, df = 9, p-value = 0.2415



In [63]:
# 수입과 직업만족도 점수 입력
income_point = matrix(c(3, 10, 20, 35,
                        3, 10, 20, 35,
                        3, 10, 20, 35,
                        3, 10, 20, 35), nrow=4, byrow=FALSE)
good_point = matrix(c(1, 3, 4, 5,
                      1, 3, 4, 5,
                      1, 3, 4, 5,
                      1, 3, 4, 5), nrow=4, byrow=TRUE)

# 점수가 반영된 데이터 생성
data2 = data * good_point;

# 행열에 이름 부여
colnames(data2) = c("매우불만족(1)", "약간만족(3)", "적절히만족(4)", "매우만족(5)")
rownames(data2) = c("<5000(3)", "5000~15000(10)", "15000~25000(20)", ">25000(35)")

# 변환 전후 데이터 비교
data
data2

매우불만족 약간만족 적절히만족 매우만족
<5000       2          4        13         3       
5000~15000  2          6        22         4       
15000~25000 0          1        15         8       
>25000      0          3        13         8

매우불만족(1) 약간만족(3) 적절히만족(4) 매우만족(5)
<5000(3)        2             12          52            15         
5000~15000(10)  2             18          88            20         
15000~25000(20) 0              3          60            40         
>25000(35)      0              9          52            40

< 변수들을 질적인 형태로 간주하기 >
수입에 대해서는 (3, 10, 20, 35), 직업만족도에 대해서는 (1, 3, 4, 5) 점수를 사용하여 변환한 데이터는 원래 데이터와 상당히 다른 형태를 보인다.

In [64]:
# 점수가 반영된 데이터로 검정
result2 = chisq.test(data2)
result2

Warning message in chisq.test(data2):
"Chi-squared approximation may be incorrect"



	Pearson's Chi-squared test

data:  data2
X-squared = 35.342, df = 9, p-value = 5.184e-05


< 검정 결과가 a와 다른 이유 >
데이터에 점수를 반영하면 데이터는 단순히 범주형이 아니라 순서형 정보를 포함하게 된다.
점수를 반영한 데이터는 변수의 순서와 강도를 추가로 반영하게 되며, 이는 검정통계량과 p값에 큰 영향을 미치며, 이에 따라 변수간의 관계에도 영향을 미친다.

In [46]:
# 수입과 직업만족도 점수를 반영하여 데이터 입력
data2 = data.frame(income = c(3, 3, 3, 3, 10, 10, 10, 10, 20, 20, 20, 20, 35, 35, 35, 35),
                   good = c(1, 3, 4, 5, 1, 3, 4, 5, 1, 3, 4, 5, 1, 3, 4, 5),
                   count = c(2, 4, 13, 3, 2, 6, 22, 4, 0, 1, 15, 8, 0, 3, 13, 8))

# 교차표 생성
data2_table = xtabs(count ~ income + good, data=data2)                   
data2_table

# 카이제곱 검정 수행
result2 = chisq.test(data2_table)
result2

      good
income  1  3  4  5
    3   2  4 13  3
    10  2  6 22  4
    20  0  1 15  8
    35  0  3 13  8

Warning message in chisq.test(data2_table):
"Chi-squared approximation may be incorrect"



	Pearson's Chi-squared test

data:  data2_table
X-squared = 11.524, df = 9, p-value = 0.2415


In [30]:
# 점수가 반영된 이차원 분할표
point = income_point * good_point
data3 = data2 * point
data3_matrix = matrix(data3, nrow=4)

# 행열에 이름 부여
colnames(data3_matrix) = c("매우불만족(1)", "약간만족(3)", "적절히만족(4)", "매우만족(5)")
rownames(data3_matrix) = c("<5000(3)", "5000~15000(10)", "15000~25000(20)", ">25000(35)")

# 비교하기
data
data3_matrix


[1]  2  2  0  0  4  6  1  3 13 22 15 13  3  4  8  8

[,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]
[1,] 2    2    0    0    4    6    1    3    13   22    15    13    3     4    
     [,15] [,16]
[1,] 8     8

매우불만족 약간만족 적절히만족 매우만족
<5000       2          4        13         3       
5000~15000  2          6        22         4       
15000~25000 0          1        15         8       
>25000      0          3        13         8

매우불만족(1) 약간만족(3) 적절히만족(4) 매우만족(5)
<5000(3)         6             12           39             9       
5000~15000(10)  60            180          660           120       
15000~25000(20)  0             80         1200           640       
>25000(35)       0            525         2275          1400

In [44]:
# 데이터 입력
data <- data.frame(
  malform = factor(c(1, 1, 1, 1, 1, 2, 2, 2, 2, 2)),
  alcohol = factor(c(0, 0.5, 1.5, 4.0, 7.0, 0, 0.5, 1.5, 4.0, 7.0)),
  count = c(17066, 14464, 788, 126, 37, 48, 38, 5, 1, 1)
)

# 교차표 생성
table_data <- xtabs(count ~ malform + alcohol, data = data)
table_data

# 카이제곱 검정 수행
chisq_test <- chisq.test(table_data)

# 결과 출력
print(chisq_test)


       alcohol
malform     0   0.5   1.5     4     7
      1 17066 14464   788   126    37
      2    48    38     5     1     1

Warning message in chisq.test(table_data):
"Chi-squared approximation may be incorrect"



	Pearson's Chi-squared test

data:  table_data
X-squared = 12.082, df = 4, p-value = 0.01675



In [48]:
# 수입과 직업만족도 점수 입력
income_point = matrix(c(3, 10, 20, 35,
                        3, 10, 20, 35,
                        3, 10, 20, 35,
                        3, 10, 20, 35), nrow=4, byrow=FALSE)
good_point = matrix(c(1, 3, 4, 5,
                      1, 3, 4, 5,
                      1, 3, 4, 5,
                      1, 3, 4, 5), nrow=4, byrow=TRUE)

# 점수가 반영된 데이터 생성
point = income_point * good_point;
data2 = data * point;

point

# 행열에 이름 부여
colnames(data2) = c("매우불만족(1)", "약간만족(3)", "적절히만족(4)", "매우만족(5)")
rownames(data2) = c("<5000(3)", "5000~15000(10)", "15000~25000(20)", ">25000(35)")

# 변환 전후 데이터 비교
data
data2

# 점수가 반영된 데이터로 검정
result2 = chisq.test(data2)
result2

Warning message in Ops.factor(left, right):
"'*' not meaningful for factors"
Warning message in Ops.factor(left, right):
"'*' not meaningful for factors"


[,1] [,2] [,3] [,4]
[1,]  3     9   12   15 
[2,] 10    30   40   50 
[3,] 20    60   80  100 
[4,] 35   105  140  175

ERROR: Error in names(x) <- value: 'names' attribute [4] must be the same length as the vector [3]


In [55]:
# 데이터 생성 (예제 데이터로 무작위 생성)
set.seed(123)  # 재현성을 위해 시드 설정
n <- 100  # 샘플 크기
X <- sample(c("A", "B", "C", "D"), n, replace = TRUE)
Y <- sample(c("W", "X", "Y", "Z"), n, replace = TRUE)

# 교차표 생성
cross_table <- table(X, Y)

# 카이제곱 독립성 검정 수행
chi_sq_test <- chisq.test(cross_table)

# 결과 출력
print(cross_table)
print(chi_sq_test)


Warning message in chisq.test(cross_table):
"Chi-squared approximation may be incorrect"


   Y
X    W  X  Y  Z
  A  3  5 13  7
  B  7  6  7  6
  C  6 12  4  7
  D  1  6  5  5

	Pearson's Chi-squared test

data:  cross_table
X-squared = 12.367, df = 9, p-value = 0.1934



In [59]:
# 필요한 패키지 설치 및 로드
if (!require(nnet)) install.packages("nnet", dependencies = TRUE)
library(nnet)

# 데이터 생성
set.seed(123)  # 재현성을 위해 시드 설정

# 독립변수 X의 수치값
X_values <- c(3, 10, 20, 35)
X <- sample(X_values, 100, replace = TRUE)  # 샘플 크기 100

# 반응변수 Y의 수준을 무작위로 생성 (예제용)
Y_levels <- c("W", "X", "Y", "Z")
Y <- sample(Y_levels, 100, replace = TRUE)

# 데이터 프레임 생성
data <- data.frame(X = X, Y = factor(Y, levels = Y_levels))

# 다중 로지스틱 회귀 분석 수행 (반응변수 Y가 범주형일 경우)
model <- multinom(Y ~ X, data = data)

# 결과 출력
summary(model)


Loading required package: nnet



# weights:  12 (6 variable)
initial  value 138.629436 
iter  10 value 134.531091
final  value 134.531081 
converged


Call:
multinom(formula = Y ~ X, data = data)

Coefficients:
  (Intercept)           X
X -0.02483502  0.03523962
Y  0.68720144 -0.01164528
Z  0.12330082  0.01775096

Std. Errors:
  (Intercept)          X
X   0.5312258 0.02838539
Y   0.5003721 0.02979744
Z   0.5313186 0.02938176

Residual Deviance: 269.0622 
AIC: 281.0622 

In [72]:
# 데이터 입력
data <- data.frame(
  malform = factor(c(1, 1, 1, 1, 1, 2, 2, 2, 2, 2)),
  alcohol = factor(c(0, 0.5, 1.5, 4.0, 7.0, 0, 0.5, 1.5, 4.0, 7.0)),
  count = c(17066, 14464, 788, 126, 37, 48, 38, 5, 1, 1)
)

# 교차표 생성
table_data <- xtabs(count ~ malform + alcohol, data = data)

# 카이제곱 검정
chi_sq_test <- chisq.test(table_data)
print(chi_sq_test)

# 선형 회귀 분석
model <- glm(count ~ malform * alcohol, data = data, family = poisson())
summary(model)


Warning message in chisq.test(table_data):
"Chi-squared approximation may be incorrect"



	Pearson's Chi-squared test

data:  table_data
X-squared = 12.082, df = 4, p-value = 0.01675




Call:
glm(formula = count ~ malform * alcohol, family = poisson(), 
    data = data)

Coefficients:
                     Estimate Std. Error  z value Pr(>|z|)    
(Intercept)          9.744843   0.007655 1273.036   <2e-16 ***
malform2            -5.873642   0.144540  -40.637   <2e-16 ***
alcohol0.5          -0.165425   0.011302  -14.637   <2e-16 ***
alcohol1.5          -3.075345   0.036437  -84.402   <2e-16 ***
alcohol4            -4.908562   0.089415  -54.896   <2e-16 ***
alcohol7            -6.133926   0.164577  -37.271   <2e-16 ***
malform2:alcohol0.5 -0.068189   0.217432   -0.314   0.7538    
malform2:alcohol1.5  0.813582   0.471340    1.726   0.0843 .  
malform2:alcohol4    1.037361   1.014307    1.023   0.3064    
malform2:alcohol7    2.262725   1.023674    2.210   0.0271 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance:  9.5424e+04  on 9  degrees of freedom
Residual deviance: -1.55

In [76]:
# 필요한 패키지 로드
if (!requireNamespace("stats", quietly = TRUE)) {
  install.packages("stats")
}
library(stats)

# 데이터 입력
data <- data.frame(
  income = factor(c(3, 3, 3, 3, 10, 10, 10, 10, 20, 20, 20, 20, 35, 35, 35, 35)),
  good = factor(c(1, 3, 4, 5, 1, 3, 4, 5, 1, 3, 4, 5, 1, 3, 4, 5)),
  count = c(2, 4, 13, 3, 2, 6, 22, 4, 0, 1, 15, 8, 0, 3, 13, 8)
)

# 독립 변수와 반응 변수의 수준을 확인
income_levels <- levels(data$income)
good_levels <- levels(data$good)

# 2x2xK 배열 생성
n_levels_income <- length(income_levels)
n_levels_good <- length(good_levels)

# 2x2 배열을 담을 3차원 배열 생성
array_data <- array(0, dim = c(n_levels_good, n_levels_good, n_levels_income),
                    dimnames = list(good_levels, good_levels, income_levels))

# 배열에 데이터 입력
for (inc in income_levels) {
  temp <- data[data$income == inc, ]
  for (i in 1:n_levels_good) {
    for (j in 1:n_levels_good) {
      # good_levels의 각 조합에 대한 count 값을 배열에 저장
      array_data[i, j, which(income_levels == inc)] <- sum(temp$count[temp$good == good_levels[i]])
    }
  }
}

# Mantel-Haenszel 검정 수행
mantelhaen_results <- mantelhaen.test(array_data)
print(mantelhaen_results)



	Cochran-Mantel-Haenszel test

data:  array_data
Cochran-Mantel-Haenszel M^2 = 0, df = 9, p-value = 1

